### PR TITLE
[1.20.5] Patch in `MutableDataComponentHolder` interface (Updates for #817)

### PR DESCRIPTION
**Pull request by ApexModder**
Copy of #817 with requested code changes

--- 

> Patches in mutable DataComponentHolder interface to allow instance checking for mutability, vanilla interface only allows for immutable lookups.
> 
> This new interface is implemented onto ItemStack and FluidStack and comes with a handy copyInto method to easily copy components from 1 holder into another.
> 
> Update: This pr includes all changes from https://github.com/neoforged/NeoForge/pull/816

### CODE REVIEW

**Review by shartte — APPROVED** [review]

File: src/main/java/net/neoforged/neoforge/common/MutableDataComponentHolder.java [hunk 1]
```diff
@@ -74,16 +74,20 @@ default <T> T remove(Supplier<? extends DataComponentType<? extends T>> componen
...
 
     /**
-     * Copes a data component from {@code src} into {@code target}
+     * Copies a data component from {@code src}
+     * 
+     * @implNote This will clear the current component value if the requested {@code src} holder does not contain a matching {@code componentType} value.
      */
-    static <T> void copyFrom(DataComponentHolder src, MutableDataComponentHolder target, DataComponentType<T> componentType) {
-        target.set(componentType, src.get(componentType));
+    default <T> void copyFrom(DataComponentHolder src, DataComponentType<T> componentType) {
```
Comment (Technici4n) [inline src/main/java/net/neoforged/neoforge/common/MutableDataComponentHolder.java]: I have a slight preference for the first argument being the `DataComponentType`, for consistency with the other methods.

**Review by Technici4n — CHANGES_REQUESTED** [review]

**Review by Technici4n — APPROVED** [review]

**Comment by ApexModder** [comment]
Added a couple more supplier methods in places where they were missing

I also noticed that when copying the mutating methods from `ItemStack` into `MutableDataComponentHolder` I missed the `applyComponents` methods, would it make sense to copy these over too or should those remain specific to the stack classes?

**Review by Technici4n — CHANGES_REQUESTED** [review]
I think that this is too much. We're adding a lot of code just to save a few `.get()` calls.

File: src/main/java/net/neoforged/neoforge/common/extensions/IDataComponentMapBuilderExtensions.java [hunk 2]
```diff
@@ -0,0 +1,21 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.common.extensions;
+
+import java.util.function.Supplier;
+import javax.annotation.Nullable;
```
Comment (Technici4n) [inline src/main/java/net/neoforged/neoforge/common/extensions/IDataComponentMapBuilderExtensions.java]: We generally use the jetbrains one.

Comment (ApexModder) [reply src/main/java/net/neoforged/neoforge/common/extensions/IDataComponentMapBuilderExtensions.java]: Whoops, so do I usually, guess idea auto filled javax for what ever reason

**Review by Technici4n — APPROVED** [review]

**Pull request merged**